File: main/3sum/3sum-fast.cs
public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums) {
        var ret = new List<IList<int>>();
        if (nums.Length < 3) return ret;

        Array.Sort(nums);
        var right_start = nums.Length - 1;
        var left_stop = right_start - 1;

        for (var left = 0; left != left_stop; ) {
            var left_elem = nums[left];
            var mid = left + 1;
            var right = right_start;

            while (mid < right) {
                var mid_elem = nums[mid];
                var right_elem = nums[right];
                var total = left_elem + mid_elem + right_elem;

                if (total == 0)
                    ret.Add(new[] { left_elem, mid_elem, right_elem });
                if (total <= 0)
                    while (++mid < right && nums[mid] == mid_elem) { }
                if (total >= 0)
                    while (mid < --right && nums[right] == right_elem) { }
            }

            while (++left != left_stop && nums[left] == left_elem) { }
        }

        return ret;
    }
}

File: main/3sum/3sum-fast.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ret = []
        if len(nums) < 3:
            return ret

        nums.sort()
        right_start = len(nums) - 1
        left_stop = right_start - 1

        left = 0
        while left != left_stop:
            left_elem = nums[left]
            mid = left + 1
            right = right_start

            while mid < right:
                mid_elem = nums[mid]
                right_elem = nums[right]
                total = left_elem + mid_elem + right_elem

                if total == 0:
                    ret.append([left_elem, mid_elem, right_elem])
                if total <= 0:
                    while mid < right and nums[mid] == mid_elem:
                        mid += 1
                if total >= 0:
                    while mid < right and nums[right] == right_elem:
                        right -= 1

            while left != left_stop and nums[left] == left_elem:
                left += 1

        return ret
